### PR TITLE
Allow re-register deleted users

### LINKED ISSUE #886: A rejected partner user is trying to request an account again
Hi @avdata99 ,

could you check on this email address in the DB, please? One partner account request has been rejected for sure, I think he should be able to request one again, with the same email...?

image

### PR DESCRIPTION
**Pull request by avdata99**
Fixes #886 

This PR stop sending an error if a user try to register with a previously registered email.
In that case, we will create a renewal request (like the one we create when a user is about to expire) with a custom message indicating user already exists an it was deleted.

image

### CODE REVIEW

File: ckanext/unhcr/blueprints/user.py [hunk 1]
```diff
@@ -137,28 +121,117 @@ def post(self):
...
+        access_request_data_dict = {
+            'object_id': deleted_user['id'],
+            'object_type': 'user',
+            'message': data_dict['message'],
+            'role': 'member',
+            'data': {
+                'default_containers': data_dict['default_containers'],
+                'user_request_type': USER_REQUEST_TYPE_RENEWAL,
+                # User xxxx {reason}
+                'reason': f'(previously deleted, rejected or expired) is trying to register again'
```
Comment (amercader) [inline ckanext/unhcr/blueprints/user.py:145]: Does this message make sense? I see it's an f-string but there are no substitutions so maybe something was lost

File: ckanext/unhcr/blueprints/user.py [hunk 2]
```diff
@@ -137,28 +121,117 @@ def post(self):
...
+
+        deleted_user = self._is_deleted_user(email)
+        if deleted_user:
+            # create the access request without creating the user
+            res = self._create_request_for_deleted_user(deleted_user, data_dict)
+        else:
+            # Creates an user and the related access request
+            res = self._create_user_and_request(context, data_dict, containers)
+        # If everithing went well, we get the user and the access request
+        # If not, we get aself.get response ready to show errors
```
Comment (amercader) [inline ckanext/unhcr/blueprints/user.py:227]: See comment above about refactoring this

**Review by amercader — COMMENTED** [review]
@avdata99 looks good overall but I suggested some code changes

File: ckanext/unhcr/blueprints/user.py [hunk 3]
```diff
@@ -137,28 +121,117 @@ def post(self):
...
+                # User xxxx {reason}
+                'reason': f'(previously deleted, rejected or expired) is trying to register again'
+            }
+        }
+        try:
+            access_request = toolkit.get_action(u'access_request_create')(
+                {'user': deleted_user['id'], 'ignore_auth': True},
+                access_request_data_dict
+            )
+        except toolkit.NotAuthorized:
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/blueprints/user.py]: @amercader I changed this [here](https://github.com/okfn/ckanext-unhcr/commit/d3a80dbcaceca0d7d8a53befaca3c777458a70dc). Now we handle this exceptions clearly.

File: ckanext/unhcr/blueprints/user.py [hunk 4]
```diff
@@ -137,28 +121,117 @@ def post(self):
...
+            toolkit.abort(403, _(u'Unauthorized to create access request %s') % u'')
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/blueprints/user.py]: Update all [here](https://github.com/okfn/ckanext-unhcr/commit/981f196ff4d8ad003ef8c66c353b48b340ba94c5)

File: ckanext/unhcr/blueprints/user.py (hunk 1 above)
Comment (avdata99) [reply ckanext/unhcr/blueprints/user.py:138]: I removed the f string. Yes, the message makes sense. It's the only difference between renewal before a user expires and a deleted user trying to access again

**Comment by avdata99** [comment]
@amercader I updated this PR because I realized we need to send an email to user if they ask for reactivation of their deleted account. To achieving that, we need to know if a renewal request comes from a deleted user or not.
Also, more tests were added.

Also applied all changes from you review
This is ready to review again.

**Review by amercader — APPROVED** [review]
Looks good @avdata99

**Pull request merged**